### ISSUE TITLE
Underline cannot be disabled

### ISSUE THREAD

**Issue by yephny** [florisboard/florisboard#1911]
#### Short description
The characters before the cursor until the beginning of the word the cursor is on are always underlined, which wasn't the case before version 0.3.16. I could not find a way to disable it in settings. Smartbar, suggestions, spell checker, and user dictionaries are turned off.

#### Steps to reproduce
1. Update Florisboard to 0.3.16 from 0.3.15.
2. Click on a text field.
3. Start typing.
4. See the underline.

#### Environment information
- FlorisBoard Version: 0.13.16
- Install Source: F-Droid
- Android: 10, LineageOS
<img src="https://a.pomf.cat/fblmtx.png" width="250" />

**Comment by patrickgold** [comment]
That's true that it cannot be disabled, because the keyboard and its input cache use this "underline" (or correctly called "composing region") to know what the current word is. This has been introduced with 0.3.16 together with its input caching for reducing stress on the slow input connection to the app.

The problem is if I add the ability to disable the composing word functionality, it confuses the cache logic and I need to add conditional statements to many actions, which can and will lead to more inconsistency. Of course atm it makes little sense without word suggestions, but once I implement them in the next milestone 0.4 it will make a lot more sense and there won't be the need to turn the underline off anymore.

**Comment by itisafrickinhighlander** [comment]
@patrickgold I am afraid it will be needed to be figured out as currently it also breaks underlining from spell-checker. Only after I add a whitespace after a word, the spell-checked underlines will show up and as soon as I start typing a word the spell-checker underlines are gone.

**Comment by yephny** [comment]
I can see how it can be vital to get a better performance out of the app, but it really does stick out, at least to my eyes. If you have noticed my custom theme, I prefer everything to be as simple and plain as possible, where nothing sticks out to distract me or even get my attention. The line is unfortunately kind of annoying and gets in the way. The 4.0 update will make no difference with regards to it for people like me, who'd rather not use word suggestions at all.

**Comment by prozacfield** [comment]
I am considering ditching this app because of this bug. As an end-user I shouldn't be bothered by your development desicions.

**Comment by tsiflimagas** [comment]
> I am considering ditching this app because of this bug. As an end-user I shouldn't be bothered by your development desicions.

Humm, an app is a sum of one's development decisions, at least the way I see it. And it goes like, when those decisions doesn't suit your needs or requirements, you look for something that does, as you said, or hope for your proposals to be accepted, thus wished changes to be made.

**Comment by Angenios** [comment]
> I am considering ditching this app because of this bug. As an end-user I shouldn't be bothered by your development desicions.

Yeah, me too, I don't understand why one day I should wake up with this update that ignores my decision not to use any correction algorithm at all, let alone underline markdown. For the moment I'll try to revert to the last update from F-Droid, then I'll search for an alternative.

**Comment by patrickgold** [comment]
I mean my development decisions will affect you and that's part of me doing software development, else I would not be allowed to change any behavior in this app ever again...

But as this rather minor change seems to be such a drastic issue for both of you I can try to add a toggle in if I find a way of doing so without creating a mess, or just tie it to the word suggestion toggle state. Will post an update here if I have a solution for this or know more.

> Unfortunately this feature also breaks word underlining from spell-checker. Only after I add a whitespace after a word, the spell-checked underlines will show up and as soon as I start typing a word the spell-checker underlines are gone.

@itisafrickinhighlander This is not indented though and a bug, on my testing devices the red spell checker underlines stay. Does this issue occur system-wide or in specific apps only?

**Comment by itisafrickinhighlander** [comment]
> @itisafrickinhighlander This is not indented though and a bug, on my testing devices the red spell checker underlines stay. Does this issue occur system-wide or in specific apps only?

I have tried some various scenarios and you are correct and I can confirm, it is app-specific, and actually from some 8 apps only Discord is affected.

**Comment by patrickgold** [comment]
> I have tried some various scenarios and you are correct and I can confirm, it is app-specific, and actually from some 8 apps only Discord is affected.

That's for the hint, just tried it out in Discord and indeed, the spell check underlines are gone. However this behavior is also reproducible with Gboard/SwiftKey/OpenBoard + any spell check service (Google and FlorisBoard), so I think this bug is on Discord's side, not mine.

**Comment by itisafrickinhighlander** [comment]
> > I have tried some various scenarios and you are correct and I can confirm, it is app-specific, and actually from some 8 apps only Discord is affected.
> 
> That's for the hint, just tried it out in Discord and indeed, the spell check underlines are gone. However this behavior is also reproducible with Gboard/SwiftKey/OpenBoard + any spell check service (Google and FlorisBoard), so I think this bug is on Discord's side, not mine.

If I may, what about hiding the underlines visually only? Would that be possible or are there some ramifications to it?

**Comment by patrickgold** [comment]
> If I may, what about hiding the underlines visually only? Would that be possible or are there some ramifications to it?

That's what I will indirectly do. Locally I already have a fix for this which basically duplicates the calculated composing range, where one is always used internally and the other is only set to something valid if the suggestions toggle is enabled. This luckily does not involve too much changing things and it looks promising.

Currently I am checking for possible inconsistencies/bugs I may create with this fix and address those, then I will push the commit and publish it in the suggestions UI logic prep PR.

**Comment by itisafrickinhighlander** [comment]
That sounds good, thank you for look into this.

**Comment by patrickgold** [comment]
Above linked PR is a fix for the underline problem. What I now basically did is to tie the visibility and public functionality of the composing region (which causes the underline) to the suggestion enable toggle:

https://user-images.githubusercontent.com/19412843/174501745-f658bd0c-2d8f-4676-84a7-0a82e5fc0e89.mp4

I hope that this satisfies everyone who saw this change as a very annoying thing, else let me know. Will be published with the next alpha release 0.4.0-alpha01.

**Comment by yephny** [comment]
@patrickgold Thanks, then I should be looking forward to 0.4.0 stable and not a hotfix.

**Comment by Goscomat** [comment]
Thanks for taking the effort to make this option available, @patrickgold. I also won't be using word suggestions, but I love using Florisboard. So this is a nice quality of life setting for people like me and @yephny :)

**Comment by yephny** [comment]
The underline is still present.

**Comment by yephny** [comment]
Any updates?

**Comment by lm41** [comment]
Do you have emoji suggestions turned on? `Emojis -> Enable emoji suggestions`

**Comment by yephny** [comment]
That was it, thanks @lm41
